src/odemis/driver/smarpod.py

```diff
@@ -279,9 +279,256 @@ def __init__(self, error_code):
         super(SmarPodError, self).__init__("Error %d. %s" % (error_code, SmarPodDLL.err_code.get(error_code, "")))
 
 
+class SA_MC_EventData(Union):
+    """
+    SA_MC event data is stored as this type of union (A C union used by DLL)
+    """
+    _fields_ = [
+         ("i32", c_int32),
+         ("i64", c_int64),
+         ("reserved", c_int8 * 32),
+         ]
+
+
+class SA_MC_Event(Structure):
+    """
+    SA_MC Event structure (C struct used by DLL)
+    """
+    _anonymous_ = ("u",)
+    _fields_ = [
+        ("type", c_uint32),
+        ("unused", c_int8 * 28),
+        ("u", SA_MC_EventData),
+        ]
+
+class SA_MC_Pose(Structure):
+    """
+    SA_MC Pose Structure (C Struct used by DLL)
+
+    Note: internally, the system uses metres and degrees for rotation
+    """
+    _fields_ = [
+        ("x", c_double),
+        ("y", c_double),
+        ("z", c_double),
+        ("rx", c_double),
+        ("ry", c_double),
+        ("rz", c_double),
+        ]
+
+    def __add__(self, o):
+        pose = Pose()
+        pose.x = self.x + o.x
+        pose.y = self.y + o.y
+        pose.z = self.z + o.z
+        pose.rx = self.rx + o.rx
+        pose.ry = self.ry + o.ry
+        pose.rz = self.rz + o.rz
+        return pose
+
+    def __sub__(self, o):
+        pose = Pose()
+        pose.x = self.x - o.x
+        pose.y = self.y - o.y
+        pose.z = self.z - o.z
+        pose.rx = self.rx - o.rx
+        pose.ry = self.ry - o.ry
+        pose.rz = self.rz - o.rz
+        return pose
+
+
+def tg_pose_to_dict(pose):
+    """
+    Convert a SA_MC Pose (C structure) to a coordinate dictionary (str) -> (double)
+        pose: a Pose (C struct)
+    returns: Coordinate dictionary (str) -> (double) of axis name to value
+    """
+    pos = {}
+    pos['x'] = pose.x
+    pos['y'] = pose.y
+    pos['z'] = pose.z
+
+    # Note: internally, the system uses metres and degrees for rotation
+    pos['rx'] = math.radians(pose.rx)
+    pos['ry'] = math.radians(pose.ry)
+    pos['rz'] = math.radians(pose.rz)
+    return pos
+
+
+def dict_to_tg_pose(pos):
+    """
+    Convert a coordinate dictionary (str) -> (double) to a SA_MC Pose C struct
+        pos: Coordinate dictionary (str) -> (double) of axis name to value
+    returns: a Pose (C struct)
+    raises ValueError if an unsupported axis name is input
+    """
+    pose = SA_MC_Pose()
+
+    # Note: internally, the system uses metres and degrees for rotation
+    for an, v in pos.items():
+        if an == "x":
```

Comment:
couldn't you rewrite this whole for loop to:
```
pose.x = pos["x"]
pose.y = pos["y"]
pose.z = pos["z"]
pose.rx = math.degrees(pos["rx"])
pose.ry = math.degrees(pos["ry"])
pose.rz = math.degrees(pos["rz"])
```

Reply:
I think I did it this way because it is possible to have a pos dict with not all of the axes i.e. pos = {'x': 0.5}

Reply:
What happens to the axes where the pose is not defined? Do they move to a random position? Or back to 0? That's all sounds fishy...

Reply:
I think it will move to 0 as that would be the value initialized. This could be problematic. Maybe this needs to be accounted for?